### PR TITLE
ADM-717[backend] Verify github and obtain github data with new API

### PR DESCRIPTION
**Pull request by BoBoDai**
## Summary

...

## Before

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

**Review by weiraneve — APPROVED** [review]

File: backend/src/main/java/heartbeat/controller/source/GithubController.java [hunk 1]
```diff
@@ -36,4 +42,28 @@ public GitHubResponse getRepos(@RequestBody @Valid SourceControlDTO sourceContro
...
 
+	@PostMapping("/{sourceType}/verify")
+	public ResponseEntity<Void> verifyToken(@PathVariable @NotBlank String sourceType,
+			@RequestBody @Valid SourceControlDTO sourceControlDTO) {
+		if (!SourceTypeEnum.isValidType(sourceType)) {
+			throw new BadRequestException("Source type is incorrect.");
+		}
+		log.info("Start to verify token");
+		gitHubService.verifyTokenV2(sourceControlDTO.getToken());
+		log.info("Successfully to verify token");
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/source/GithubController.java:53]: remove `to`

File: backend/src/main/java/heartbeat/controller/source/GithubController.java [hunk 2]
```diff
@@ -36,4 +42,28 @@ public GitHubResponse getRepos(@RequestBody @Valid SourceControlDTO sourceContro
...
+		return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
+	}
+
+	@PostMapping("/{sourceType}/repos/branches/{branch}/verify")
+	public ResponseEntity<Void> verifyBranch(@PathVariable @NotBlank String sourceType,
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/source/GithubController.java:58]: `sourceType` should be a enum

File: backend/src/main/java/heartbeat/controller/source/GithubController.java [hunk 3]
```diff
@@ -36,4 +42,28 @@ public GitHubResponse getRepos(@RequestBody @Valid SourceControlDTO sourceContro
...
+			@PathVariable @NotBlank String branch, @RequestBody @Valid VerifyBranchRequest request) {
+		if (!SourceTypeEnum.isValidType(sourceType)) {
+			throw new BadRequestException("Source type is incorrect.");
+		}
+		log.info("Start to verify target branch.");
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/source/GithubController.java:63]: start ... logged `sourceType` and `branch`

File: backend/src/main/java/heartbeat/controller/source/dto/VerifyBranchRequest.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+@Getter
+@Setter
+public class VerifyBranchRequest {
+
+	@NotNull(message = "Token cannot be empty.")
+	@Pattern(regexp = "^(ghp|gho|ghu|ghs|ghr)_([a-zA-Z0-9]{36})$", message = "token's pattern is incorrect")
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/source/dto/VerifyBranchRequest.java:20]: extract `"^(ghp|gho|ghu|ghs|ghr)_([a-zA-Z0-9]{36})$",` into a constant, and shared with `SourceControlDTO`

File: backend/src/main/java/heartbeat/controller/source/GithubController.java (hunk 2 above)
Comment (guzhongren) [reply backend/src/main/java/heartbeat/controller/source/GithubController.java:58]: BTW, `sourceType` value should use lowercase like `buildkite`, `github`

File: backend/src/main/java/heartbeat/service/source/github/GitHubService.java [hunk 5]
```diff
@@ -49,6 +50,7 @@ public void shutdownExecutor() {
 		customTaskExecutor.shutdown();
 	}
 
+	@Deprecated
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/source/github/GitHubService.java:53]: when to deprecated? please give more context or next step description

File: backend/src/main/java/heartbeat/service/source/github/GitHubService.java [hunk 6]
```diff
@@ -87,6 +89,46 @@ public GitHubResponse verifyToken(String githubToken) {
 		}
 	}
 
+	public void verifyTokenV2(String githubToken) {
+		try {
+			String token = "token " + githubToken;
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/source/github/GitHubService.java:94]: extract "token "  as a constant, shared with `verifyToken`

File: backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java [hunk 7]
```diff
@@ -57,18 +62,175 @@ void shouldReturnOkStatusAndCorrectResponseWithRepos() throws Exception {
 	}
 
 	@Test
+	void shouldReturnNoContentStatusWhenVerifyToken() throws Exception {
+		doNothing().when(gitHubVerifyService).verifyTokenV2(GITHUB_TOKEN);
+		SourceControlDTO sourceControlDTO = SourceControlDTO.builder().token(GITHUB_TOKEN).build();
+
+		mockMvc
+			.perform(post("/source-control/GitHub/verify")
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java:70]: GitHub should not use upppercase name in the URL path!!

File: backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java [hunk 8]
```diff
@@ -57,18 +62,175 @@ void shouldReturnOkStatusAndCorrectResponseWithRepos() throws Exception {
...
+	@Test
+	void shouldReturnNoContentStatusWhenVerifyTargetBranch() throws Exception {
+		VerifyBranchRequest verifyBranchRequest = VerifyBranchRequest.builder()
+			.repository(GITHUB_REPOSITORY)
+			.token(GITHUB_TOKEN)
+			.build();
+		doNothing().when(gitHubVerifyService).verifyCanReadTargetBranch("fake/repo", "main", GITHUB_TOKEN);
+
+		mockMvc
+			.perform(post("/source-control/GitHub/repos/branches/main/verify")
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java:85]: GitHub

File: backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java [hunk 9]
```diff
@@ -57,18 +62,175 @@ void shouldReturnOkStatusAndCorrectResponseWithRepos() throws Exception {
...
+		final var result = JsonPath.parse(content).read("$.token").toString();
+		assertThat(result).contains("Token cannot be empty.");
+	}
+
+	@Test
+	void shouldReturnBadRequestGivenRequestBodyIsNullWhenVerifyToken() throws Exception {
+		SourceControlDTO sourceControlDTO = SourceControlDTO.builder().build();
+
+		final var response = mockMvc
+			.perform(post("/source-control/GitHub/verify")
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java:113]: GitHub

File: backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java [hunk 10]
```diff
@@ -57,18 +62,175 @@ void shouldReturnOkStatusAndCorrectResponseWithRepos() throws Exception {
...
+		final var result = JsonPath.parse(content).read("$.token").toString();
+		assertThat(result).contains("Token cannot be empty.");
+	}
+
+	@Test
+	void shouldReturnBadRequestGivenRequestBodyIsNullWhenVerifyBranch() throws Exception {
+		VerifyBranchRequest verifyBranchRequest = VerifyBranchRequest.builder().build();
+
+		final var response = mockMvc
+			.perform(post("/source-control/GitHub/repos/branches/main/verify")
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java:130]: GitHub

File: backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java [hunk 11]
```diff
@@ -57,18 +62,175 @@ void shouldReturnOkStatusAndCorrectResponseWithRepos() throws Exception {
...
+		final var result = JsonPath.parse(content).read("$.token").toString();
+		assertThat(result).contains("Token cannot be empty.");
+	}
+
+	@Test
+	void shouldReturnBadRequestGivenRepositoryIsNullWhenVerifyBranch() throws Exception {
+		VerifyBranchRequest verifyBranchRequest = VerifyBranchRequest.builder().token(GITHUB_TOKEN).build();
+
+		final var response = mockMvc
+			.perform(post("/source-control/GitHub/repos/branches/main/verify")
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java:147]: GitHub

File: backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java [hunk 12]
```diff
@@ -57,18 +62,175 @@ void shouldReturnOkStatusAndCorrectResponseWithRepos() throws Exception {
...
+		final var result = JsonPath.parse(content).read("$.repository").toString();
+		assertThat(result).contains("Repository is required.");
+	}
+
+	@Test
+	void shouldReturnBadRequestGivenSourceTypeIsWrongWhenVerifyToken() throws Exception {
+		SourceControlDTO sourceControlDTO = SourceControlDTO.builder().token(GITHUB_TOKEN).build();
+
+		final var response = mockMvc
+			.perform(post("/source-control/github/verify")
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/controller/source/GithubControllerTest.java:164]: Goood

File: backend/src/test/java/heartbeat/service/source/github/GithubServiceTest.java [hunk 13]
```diff
@@ -180,6 +189,28 @@ void shouldReturnNonRedundantGithubReposWhenCallGithubFeignClientApi() {
 	}
 
 	@Test
+	void shouldReturnGithubTokenIsVerifyWhenVerifyToken() {
+		String githubToken = "123456";
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/source/github/GithubServiceTest.java:193]: `githubToken` should use `GITHUB_TOKEN` in the testFixtures.java

same as others...

File: backend/src/test/java/heartbeat/service/source/github/GithubServiceTest.java [hunk 14]
```diff
@@ -180,6 +189,28 @@ void shouldReturnNonRedundantGithubReposWhenCallGithubFeignClientApi() {
...
+	}
+
+	@Test
+	void shouldReturnGithubBranchIsVerifyWhenVerifyBranch() {
+		String githubToken = "123456";
+		String token = "token " + githubToken;
+
+		doNothing().when(gitHubFeignClient).verifyCanReadTargetBranch(any(), any(), any());
+
+		assertDoesNotThrow(() -> githubService.verifyCanReadTargetBranch(GITHUB_REPOSITORY, "main", githubToken));
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/source/github/GithubServiceTest.java:208]: assertDoesNotThrow is not necessary,
given `new line` between `when` and then

File: backend/src/test/java/heartbeat/service/source/github/GithubServiceTest.java [hunk 15]
```diff
@@ -215,6 +247,59 @@ void shouldGithubReturnEmptyWhenVerifyGithubThrowGithubRepoEmptyException() {
...
+			.verifyCanReadTargetBranch("fake/repo", "main", "token " + githubEmptyToken);
+
+		var exception = assertThrows(UnauthorizedException.class,
+				() -> githubService.verifyCanReadTargetBranch(GITHUB_REPOSITORY, "main", githubEmptyToken));
+		assertEquals("Failed to get GitHub info_status: 401 UNAUTHORIZED, reason: ...", exception.getMessage());
+	}
+
+	@Test
+	void shouldThrowExceptionGivenGithubReturnPermissionDenyExceptionWhenVerifyBranch() {
+		String githubEmptyToken = "123456";
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/source/github/GithubServiceTest.java:275]: duplcated code

File: backend/src/test/java/heartbeat/service/source/github/GithubServiceTest.java [hunk 16]
```diff
@@ -215,6 +247,59 @@ void shouldGithubReturnEmptyWhenVerifyGithubThrowGithubRepoEmptyException() {
...
+		assertEquals("Unable to read target branch", exception.getMessage());
+	}
+
+	@Test
+	void shouldThrowExceptionGivenGithubReturnCompletionExceptionExceptionWhenVerifyToken() {
+		String githubEmptyToken = "123456";
+		doThrow(new CompletionException(new Exception("UnExpected Exception"))).when(gitHubFeignClient)
+			.verifyToken("token " + githubEmptyToken);
+
+		assertThrows(InternalServerErrorException.class, () -> githubService.verifyTokenV2(githubEmptyToken));
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/source/github/GithubServiceTest.java:290]: any other assertion for the exception content or message

File: backend/src/test/java/heartbeat/service/source/github/GithubServiceTest.java [hunk 17]
```diff
@@ -215,6 +247,59 @@ void shouldGithubReturnEmptyWhenVerifyGithubThrowGithubRepoEmptyException() {
...
+	}
+
+	@Test
+	void shouldThrowExceptionGivenGithubReturnUnauthorizedExceptionWhenVerifyBranch() {
+		String githubEmptyToken = "123456";
+		doThrow(new CompletionException(new Exception("UnExpected Exception"))).when(gitHubFeignClient)
+			.verifyCanReadTargetBranch("fake/repo", "main", "token " + githubEmptyToken);
+
+		assertThrows(InternalServerErrorException.class,
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/source/github/GithubServiceTest.java:299]: any other assertion for the exception content or message?

**Review by guzhongren — COMMENTED** [review]
Change with the comments

**Pull request merged**